contributors/design-proposals/cpu-affinity-numa-awareness.md

```diff
@@ -0,0 +1,216 @@
+
+In order to support NUMA affined workloads, the `Node` must make its
+NUMA topology available for introspection by other agents that schedule
+pods.
+
+This proposal recommends that the `NodeStatus` is augmented as follows:
+
+```
+// NodeStatus is information about the current status of a node.
+type NodeStatus struct {
+    ...
+	// Topology represents the NUMA topology of a node to aid NUMA aware scheduling.
+	// +optional
+	Topology NUMATopology
+}
+
+// NUMATopology describes the NUMA topology of a node.
+type NUMATopology struct {
+	// NUMANodes represents the list of NUMA nodes in the topology.
+	NUMANodes []NUMANode
+}
+
+// NUMANode describes a single NUMA node.
+type NUMANode struct {
+	// Identifies a NUMA node on a single host.
+	NUMANodeID string
+	// Capacity represents the total resources associated to the NUMA node.
+	// cpu: 4 <number of cores>
+	// memory: <amount of memory in normal page size>
+	// hugepages: <amount of memory in huge page size>
+	Capacity ResourceList
+	// Allocatable represents the resources of a NUMA node that are available for scheduling.
+	// +optional
+	Allocatable ResourceList
+	// CPUSet represents the physical numbers of the CPU cores
+	// associated with this node.
+	// Example: 0-3 or 0,2,4,6
+	// The values are expressed in the List Format syntax specified
+	// here: http://man7.org/linux/man-pages/man7/cpuset.7.html
+	CPUSet string
+}
+```
+
+## Node Configuration
+
+### Isolating host processes
+
+By default, load balancing is done across all CPUs, except those marked isolated
+using the kernel boot time `isolcpus=` argument.  When configuring a node to support
+CPU and NUMA affinity, many operators may wish to isolate host processes to particular
+cores.
+
+It is recommended that operators set a CPU value for `--system-reserved`
+in whole cores that aligns with the set of cpus that are made available to the default
+kernel scheduling algorithm.  If an operator is on a `systemd` managed platform, they
+may choose instead to set the `CPUAffinity` value for the root slice to the set of CPU
+cores that are reserved for the host processes.
+
+**TODO**
+
+1. how should `kubelet` discover the reserved `cpu-set` value?
+1. in a numa system, `kubelet` reservation for memory needs to be removed from
+a particular numa node capacity so numa node allocatable is as expected. 
+
+### Configuring Taints
+
+The following `Taint` keys are defined to enable CPU pinning and NUMA awareness.
+
+#### CPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `dedicated`
+
+If `dedicated`, all pods that match this taint will require dedicated compute resources. Each
+pod bound to this node must request CPU in whole cores.  The CPU limit must equal the request.
+
+#### NUMACPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `strict`
+
+If `strict`, all pods that match this taint must request CPU (whole or fractional cores) that
+fit a single NUMA node `cpu` allocatable.
```

Comment:
This feels somewhat inflexible wrt potential future expansion. It would be better to express a desired number of node count. eg  NUMANodes=N, where N is one or more, to indicate how many NUMA nodes it is willing / able to consume. eg an application inside a container may be capable of dividing its workload into 2 pieces, so could deal wtih being spread on 2 NUMA nodes, thus could request NUMANodes=1,2 to indicate its able to deal with 1 or 2 NUMA nodes, but not more than that.